swarm/src/behaviour/external_addresses.rs

```diff
@@ -29,18 +18,40 @@ impl ExternalAddresses {
         self.addresses.iter()
     }
 
+    pub fn as_slice(&self) -> &[Multiaddr] {
+        self.addresses.as_slice()
+    }
+
     /// Feed a [`FromSwarm`] event to this struct.
     ///
     /// Returns whether the event changed our set of external addresses.
     pub fn on_swarm_event<THandler>(&mut self, event: &FromSwarm<THandler>) -> bool {
         match event {
             FromSwarm::ExternalAddrConfirmed(ExternalAddrConfirmed { addr }) => {
-                if self.addresses.len() < self.limit {
-                    return self.addresses.insert((*addr).clone());
+                if self.addresses.contains(addr) {
+                    return false;
+                }
+
+                if self.addresses.len() >= MAX_LOCAL_EXTERNAL_ADDRS {
+                    // If we hit the limit, we can add the address in O(1):
```

Comment:
Just so that we are on the same page. The `self.addresses` `Vec` doesn't grow beyond `<100` addresses. Also I would expect change frequency to this datastructure on the order of `> 1s`. Thus I don't think even an `O(n^2)` would be a problem here.

Reply:
Removed.